pkg/server/updates.go

```diff
@@ -183,6 +184,152 @@ func addInfoToURL(ctx context.Context, url *url.URL, s *Server, n diagnosticspb.
 	url.RawQuery = q.Encode()
 }
 
+const (
+	aws   = "Amazon Web Services"
+	gcp   = "Google Cloud Platform"
+	azure = "Microsoft Azure"
+)
+
+// parseAWSInstanceMetadata uses the structure described
+// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+// If we encounter JSON we cannot marhsal into this structure, we
+// assume we're not running on AWS.
+func parseAWSInstanceMetadata(body []byte) (bool, string, string) {
+	instanceMetadata := struct {
+		InstanceClass string `json:"instanceType"`
+	}{}
+
+	success := true
+	if err := json.Unmarshal(body, &instanceMetadata); err != nil {
+		success = false
+	}
+
+	return success, aws, instanceMetadata.InstanceClass
+}
+
+// parseGCPInstanceMetadata relies on the structure indicated at
+// https://cloud.google.com/compute/docs/storing-retrieving-metadata
+// If we encounter a string that doesn't match our format, we  assume
+// we're not running on GCP.
+func parseGCPInstanceMetadata(body []byte) (bool, string, string) {
+	bodyStr := string(body)
+
+	// The structure of the API's response can be found at
+	// https://cloud.google.com/compute/docs/storing-retrieving-metadata;
+	// look for machine-type
+	instanceClassRE := regexp.MustCompile(`machineTypes\/(.+)$`)
+
+	instanceClass := instanceClassRE.FindStringSubmatch(bodyStr)
+
+	// Regex should only have 2 values: matched string and
+	// capture group containing the machineTypes value.
+	if len(instanceClass) != 2 {
+		return false, "", ""
+	}
+
+	return true, gcp, instanceClass[1]
+}
+
+// parseAzureInstanceMetadata uses the structure described
+// https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service
+// If we encounter JSON we cannot marhsal into this structure, we
+// assume we're not running on Azure.
+func parseAzureInstanceMetadata(body []byte) (bool, string, string) {
+	instanceMetadata := struct {
+		ComputeEnv struct {
+			InstanceClass string `json:"vmSize"`
+		} `json:"compute"`
+	}{}
+
+	success := true
+	if err := json.Unmarshal(body, &instanceMetadata); err != nil {
+		success = false
+	}
+
+	return success, azure, instanceMetadata.ComputeEnv.InstanceClass
+}
+
+type metadataReqHeader struct {
+	key   string
+	value string
+}
+
+func getInstanceMetadata(url string, headers []metadataReqHeader) ([]byte, error) {
+	client := http.Client{
+		Timeout: time.Duration(500 * time.Millisecond),
```

Comment:
Go keeps constants like `500` untyped until it has to, so it can actually figure out that `500*time.Millisecond` should break down into basically `time.Duration(500) * time.Duration(1000000)` which is still a `time.Duration` so there's no need to wrap it it in cast to `time.Duration` (and indeed, our linters will complain about the extra cast).